src/topology/extend_from_subset.lean

```diff
@@ -0,0 +1,89 @@
+
+-/
+
+noncomputable theory
+
+open_locale topological_space
+open filter set
+
+variables {X Y : Type*} [topological_space X] [topological_space Y]
+
+/-- Extend a function from a set `A`. The resulting function `g` is such that
+at any `x₀`, if `f` converges to some `y` as `x` tends to `x₀` within `A`,
+then `g x₀` is defined to be one of these `y`. Else, `g x₀` could be anything. -/
+def extend_from (A : set X) (f : X → Y) : X → Y :=
+λ x, @@lim _ ⟨f x⟩ (nhds_within x A) f
+
+/-- If `f` converges to some `y` as `x` tends to `x₀` within `A`,
+then `f` tends to `extend_from A f x` as `x` tends to `x₀`. -/
+lemma tendsto_extend_from {A : set X} {f : X → Y} {x : X}
+  (h : ∃ y, tendsto f (nhds_within x A) (𝓝 y)) : tendsto f (nhds_within x A) (𝓝 $ extend_from A f x) :=
+lim_spec h
+
+lemma extend_from_eq [t2_space Y] {A : set X} {f : X → Y} {x : X} {y : Y} (hx : x ∈ closure A)
+  (hf : tendsto f (nhds_within x A) (𝓝 y)) : extend_from A f x = y :=
+begin
+  haveI := mem_closure_iff_nhds_within_ne_bot.mp hx,
+  exact tendsto_nhds_unique (lim_spec ⟨y, hf⟩) hf,
+end
+
+lemma extend_from_extends [t2_space Y] {f : X → Y} {A : set X} (hf : continuous_on f A) :
+  ∀ x ∈ A, extend_from A f x = f x :=
+λ x x_in, extend_from_eq (subset_closure x_in) (hf x x_in)
+
+/-- If `f` is a function to a regular space `Y` which is continuous on a set `A`,
+then `extend_from A f` is continuous on `B ⊆ closure A`, provided that `f` has a limit
+within `A` at any point in `B`. -/
+lemma continuous_on_extend_from [regular_space Y] {f : X → Y} {A B : set X} (hB : B ⊆ closure A)
+  (hf : ∀ x ∈ B,  ∃ y, tendsto f (nhds_within x A) (𝓝 y)) : continuous_on (extend_from A f) B :=
+begin
+  set φ := extend_from A f,
+  intros x x_in,
+  suffices : ∀ V' ∈ 𝓝 (φ x), is_closed V' → φ ⁻¹' V' ∈ nhds_within x B,
+    by simpa [continuous_within_at, (closed_nhds_basis _).tendsto_right_iff],
+  intros V' V'_in V'_closed,
+  obtain ⟨V, V_in, V_op, hV⟩ : ∃ V ∈ 𝓝 x, is_open V ∧ V ∩ A ⊆ f ⁻¹' V',
+  { have := tendsto_extend_from (hf x  x_in),
+    rcases (nhds_within_basis_open x A).tendsto_left_iff.mp this V' V'_in with ⟨V, ⟨hxV, V_op⟩, hV⟩,
+    use [V, mem_nhds_sets V_op hxV, V_op, hV] },
+  suffices : ∀ y ∈ V ∩ B, φ y ∈ V',
+    from mem_sets_of_superset (inter_mem_inf_sets V_in $ mem_principal_self B) this,
+  rintros y ⟨hyV, hyB⟩,
+  haveI := mem_closure_iff_nhds_within_ne_bot.mp (hB hyB),
+  have limy : tendsto f (nhds_within y A) (𝓝 $ φ y) := tendsto_extend_from (hf y hyB),
+  have hVy : V ∈ 𝓝 y := mem_nhds_sets V_op hyV,
+  have : V ∩ A ∈ (nhds_within y A),
+    by simpa [inter_comm] using inter_mem_nhds_within _ hVy,
+  exact mem_of_closed_of_tendsto limy V'_closed (mem_sets_of_superset this hV)
+end
+
+/-- If a function `f` to a regular space `Y` is continuous on a dense set `A`,
+then `extend_from A f` is continuous, provided that `f` has a limit
+within `A` for any `x`. -/
+lemma continuous_extend_from [regular_space Y] {f : X → Y} {A : set X} (hA : ∀ x, x ∈ closure A)
```

Comment:
The assumption `hA` is strange. Why not `closure A = univ` or `univ ⊆ closure A`?

Reply:
I think we are inconsistent here. Because we never defined `is_dense A`, we have various ways to express this. Note that this was resolved for [maps with dense range](https://github.com/leanprover-community/mathlib/blob/1bb3d198adb92b0fe00b922a357cc29f54c35e70/src/topology/dense_embedding.lean#L31-L40), with lemmas going back and forth, but somehow this was never ported to dense subsets.

Reply:
I was indeed surprised I couldn't express this as `A` being a dense set. In the meantime, I chose to keep the weaker requirement (inclusion), but maybe that's inconsistent since most lemmas involving the world `dense` in the topological meaning use `closure A = univ` (e.g `dense_iff_inter_open`, `dense_of_subset_dense`, ...)